addons/binding/org.openhab.binding.squeezebox/src/main/java/org/openhab/binding/squeezebox/internal/SqueezeBoxStateDescriptionOptionsProvider.java

```diff
@@ -0,0 +1,60 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.squeezebox.internal;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.type.DynamicStateDescriptionProvider;
+import org.eclipse.smarthome.core.types.StateDescription;
+import org.eclipse.smarthome.core.types.StateOption;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * Dynamic provider of state options while leaving other state description fields as original.
+ *
+ * @author Gregory Moyer - Initial contribution
+ * @author Mark Hilbush - Adapted to squeezebox binding
+ */
+@Component(service = { DynamicStateDescriptionProvider.class,
+        SqueezeBoxStateDescriptionOptionsProvider.class }, immediate = true)
+public class SqueezeBoxStateDescriptionOptionsProvider implements DynamicStateDescriptionProvider {
+    private final Map<ChannelUID, List<StateOption>> channelOptionsMap = new ConcurrentHashMap<>();
+
+    public void setStateOptions(ChannelUID channelUID, List<StateOption> options) {
+        channelOptionsMap.put(channelUID, options);
+    }
+
+    @Override
+    public @Nullable StateDescription getStateDescription(@NonNull Channel channel, @Nullable StateDescription original,
```

Comment:
CC @triller-telekom if I remember correctly we should not be using both of these in this case it should 

See also https://www.eclipse.org/smarthome/documentation/development/guidelines.html#a-code-style A.12

Reply:
@triller-telekom Can you also weigh in on this?

Reply:
I can jump in here: In general we want to avoid using @NonNull but use @NonNullByDefault on class level. You can have a look at #3207 I already changed adopted it in my PR.

Reply:
@mhilbush Yes I second @cweitkamp's comment, we are only using `@NonNullByDefault` and `@Nullable`, see the link that @martinvw has posted.